cmd/collector/collector.go

```diff
@@ -0,0 +1,178 @@
+)
+
+var (
+	IPFIXAddr      string
+	IPFIXPort      uint16
+	IPFIXTransport string
+)
+
+func initLoggingToFile(fs *pflag.FlagSet) {
+	var err error
+	var logToStdErr bool
+
+	logToStdErr, err = fs.GetBool(logToStdErrFlag)
+	if err != nil {
+		// Should not happen. Return for safety.
+		return
+	}
+	if logToStdErr {
+		// Logging to files is not enabled.
+		return
+	}
+}
+
+func addIPFIXFlags(fs *pflag.FlagSet) {
+	fs.StringVar(&IPFIXAddr, "ipfix.addr", "", "IPFIX collector address")
+	fs.Uint16Var(&IPFIXPort, "ipfix.port", 4739, "IPFIX collector port")
+	fs.StringVar(&IPFIXTransport, "ipfix.transport", "tcp", "IPFIX collector transport layer")
+}
+
+func printIPFIXMessage(msg *entities.Message) {
+	var buf bytes.Buffer
+	fmt.Fprint(&buf, "\nIPFIX-HDR:\n")
+	fmt.Fprintf(&buf, "  version: %v,  Message Length: %v\n", msg.Version, msg.BufferLength)
+	fmt.Fprintf(&buf, "  Exported Time: %v (%v)\n", msg.ExportTime, time.Unix(int64(msg.ExportTime), 0))
+	fmt.Fprintf(&buf, "  Sequence No.: %v,  Observation Domain ID: %v\n", msg.SeqNumber, msg.ObsDomainID)
+
+	if msg.Set.GetSetType() == entities.Template {
+		fmt.Fprint(&buf, "TEMPLATE SET:\n")
+		for i, record := range msg.Set.GetRecords() {
+			fmt.Fprintf(&buf, "  TEMPLATE RECORD-%d:\n", i)
+			for _, ie := range record.GetInfoElements() {
+				fmt.Fprintf(&buf, "    %s: len=%d (enterprise ID = %d) \n", ie.Element.Name, ie.Element.Len, ie.Element.EnterpriseId)
+			}
+		}
+	} else {
+		fmt.Fprint(&buf, "DATA SET:\n")
+		for i, record := range msg.Set.GetRecords() {
+			fmt.Fprintf(&buf, "  DATA RECORD-%d:\n", i)
+			for _, ie := range record.GetInfoElements() {
+				fmt.Fprintf(&buf, "    %s: %v \n", ie.Element.Name, ie.Value)
+			}
+		}
+	}
+	klog.Infof(buf.String())
+}
+
+func signalHandler(signalCh chan os.Signal, messageReceived chan *entities.Message) {
+	for {
+		select {
+		case msg := <-messageReceived:
+			printIPFIXMessage(msg)
+		case <-signalCh:
+			break
+		}
+	}
+}
+
+func run() error {
+	klog.Info("Starting IPFIX collector")
+
+	// Load the IPFIX global registry
+	registry.LoadRegistry()
+
+	var netAddr net.Addr
+	var err error
+	if IPFIXTransport == "tcp" {
+		netAddr, err = net.ResolveTCPAddr("tcp", IPFIXAddr+":"+strconv.Itoa(int(IPFIXPort)))
+		if err != nil {
+			return err
+		}
+	} else if IPFIXTransport == "udp" {
+		netAddr, err = net.ResolveUDPAddr("udp", IPFIXAddr+":"+strconv.Itoa(int(IPFIXPort)))
+		if err != nil {
+			return err
+		}
+	} else {
+		return fmt.Errorf("input given ipfix.transport flag is not supported or valid")
+	}
+	// Initialize collecting process
+	cp, err := collector.InitCollectingProcess(netAddr, 65535, 0)
+	if err != nil {
+		return err
+	}
+	// Start listening to connections and receiving messages.
+	messageReceived := make(chan *entities.Message)
+	go func() {
+		go cp.Start()
+		msgCount := 0
+		// This can be simplified if collector uses a message channel.
+		// Are there any use cases for having list of messages in collector process.
```

Comment:
I'm a little confused about this comment. Do you mean there are still use cases of having a list?

Reply:
Got it. I will modify it in intermediate process PR.

Reply:
Thanks.